### DISCUSSION TITLE
Linux Brew unlink perl safe? for system caffeine (apt)

### DISCUSSION THREAD

**Discussion by yash-fn** [Homebrew/discussions#4775] [Linux] [answered]
### Additional Info

- [ ] I am using Windows Subsystem for Linux

### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.10-7-g704b97d
ORIGIN: https://github.com/Homebrew/brew
HEAD: 704b97d36d0cc8263cb0775d5b695b343145485e
Last commit: 18 hours ago
Core tap JSON: 08 Sep 21:56 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :1
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 12
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 2.6.10 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: dodeca-core 64-bit skylake
Clang: 16.0.6
Git: 2.34.1 => /bin/git
Curl: 7.81.0 => /bin/curl
Kernel: Linux 5.15.0-82-generic x86_64 GNU/Linux
OS: Ubuntu 22.04.3 LTS (jammy)
Host glibc: 2.35
/usr/bin/gcc: 11.4.0
/usr/bin/ruby: 3.0.2
glibc: N/A
gcc@11: N/A
gcc: 13.2.0
xorg: N/A
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/jbig2arith.h
  /usr/local/include/jbig2comparator.h
  /usr/local/include/jbig2segments.h
  /usr/local/include/jbig2structs.h
  /usr/local/include/jbig2sym.h

Warning: Unbrewed '.la' files were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected '.la' files:
  /usr/local/lib/libjbig2enc.la

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libjbig2enc.a

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  perl
```

### Description of issue

I need system caffeine from Ubuntu 22.04 apt sources (`apt install caffeine`) to prevent system sleep for certain tasks. System caffeine doesn't work because apparently it uses perl and it keeps trying to use homebrew perl instead of system perl. `brew unlink perl` solves all my problems, but now `brew doctor` now will complain as shown above. brew perl was installed as a dependency of vim. Is this safe? will brew-installed vim work fine with an unlinked perl? Is there a way to get a keg-only version of perl for vim?

**Comment by gromgit** [comment]
> System caffeine doesn't work because apparently it uses perl and it keeps trying to use homebrew perl instead of system perl.

That would be very weird indeed, since Caffeine is written in _Python_:
```
$ file /usr/bin/caffeine
/usr/bin/caffeine: Python script, Unicode text, UTF-8 text executable
```
What is the exact error message you're seeing?

**Reply by yash-fn** [reply]
Huh. that is weird. well here is the output:

```python
❯❯: caffeinate sleep 1
Can't locate Net/DBus.pm in @INC (you may need to install the Net::DBus module) (@INC contains: /home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/site_perl/5.36/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/site_perl/5.36 /home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/5.36/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/5.36 /home/linuxbrew/.linuxbrew/lib/perl5/site_perl/5.36/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/lib/perl5/site_perl/5.36) at -e line 6.
BEGIN failed--compilation aborted at -e line 6.
❯❯: brew unlink perl
Unlinking /home/linuxbrew/.linuxbrew/Cellar/perl/5.36.1... 2489 symlinks removed.
❯❯: caffeinate sleep 1
❯❯:
```

I saw a bunch of linuxbrew paths and what appeared to be a perl module problem so I tried unlinking perl and all my problems went away.

**Reply by gromgit** [reply]
Hmmm. It's most likely something that `caffeinate` is trying to run then, that's not part of Caffeine.

The simplest way to force whatever's launched by `caffeinate` to use system Perl, without having to unlink Homebrew Perl:
```
PATH=/usr/bin:$PATH caffeinate sleep 1
```
UPDATE: Whatever Perl stuff is being run also appears to be only on your system...
```bash
# I have Homebrew Perl installed and first in my PATH
gromgit$ which perl
/home/linuxbrew/.linuxbrew/bin/perl

# No Net::DBus module installed
gromgit$ find /home/linuxbrew/.linuxbrew/Cellar/perl/5.36.1/ -name DBus.pm

# No errors running caffeinate
gromgit$ caffeinate sleep 1

gromgit$
```

**Reply by cho-m** [reply, marked as answer]
Reason I see is that Ubuntu's `caffeinate` runs `xdg-screensaver` - https://bazaar.launchpad.net/~caffeine-developers/caffeine/main/view/head:/caffeinate#L64

That will run a perl script https://cgit.freedesktop.org/xdg/xdg-utils/tree/scripts/xdg-screensaver.in#n448

**Reply by yash-fn** [reply]
Oh interesting. Thanks to you both for your help! The environment argument fixed my issue short term. Long term, does this seem like something I should file with the Ubuntu guy's? Maybe they need to modify their software to use system perl?

**Reply by yash-fn** [reply]
Well, so I filed this issue with them at: https://bugs.launchpad.net/ubuntu/+source/caffeine/+bug/2035241.

**Reply by yash-fn** [reply]
Okay, so at line 448, as @cho-m mentioned, i changed `perl` to `/usr/bin/perl` and now I can run caffeine with homebrew perl appropriately linked! Thanks guys!!!
